rs7903146 <- read.csv("rs7903146.csv") 
rs7903146_props <- rs7903146 %>% 
  select(-c("Control_CC", "Control_CT", "Control_TT", "HWE")) %>%
  pivot_longer(cols = starts_with("T2DM"),
               names_prefix = "T2DM_",
               names_to = "Variant",
               values_to = "Count") %>%
  mutate(Ethnicity = case_when(
    Ethnicity == "Caucasian" ~ "Euro",
    Ethnicity == "Black African" ~ "Black",
    Ethnicity %in% c("East Asian", "South Asian") ~ "Asian",
    TRUE ~ "Other"
  )) %>% 
  group_by(Ethnicity, Variant) %>% 
  summarise(Total = sum(Count), .groups="drop") %>% 
  group_by(Ethnicity) %>% 
  mutate(Prop = Total / sum(Total)) %>%
  select(-Total)

examples_dir = system.file("examples", package = "sim1000G")
vcf_file = file.path(examples_dir, "region.vcf.gz")
vcf = readVCF(vcf_file, maxNumberOfVariants = 100, min_maf = 0.02, max_maf = NA )

# 17 Genotypes Variants, 
genotype_marginal <- list(
  c(0.2, 0.6, 0.2),
  c(0.1, 0.4, 0.5),
  c(0.3, 0.5, 0.2),
  c(0.3, 0.4, 0.3),
  c(0.2, 0.5, 0.3),
  c(0.3, 0.4, 0.3),
  c(0.1, 0.2, 0.7),
  c(0.2, 0.5, 0.3),
  c(0.3, 0.5, 0.2),
  c(0.2, 0.6, 0.2),
  c(0.3, 0.5, 0.2),
  c(0.2, 0.3, 0.5),
  c(0.1, 0.5, 0.4),
  c(0.2, 0.3, 0.5),
  c(0.1, 0.2, 0.7),
  c(0.2, 0.5, 0.3),
  c(0.1, 0.6, 0.3)
)
genotype_correlation <- rbind(
  c(1.000, 0.000, 0.000, 0.107, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.119, 0.000, 0.000, 0.000, 0.000, 0.000), # G_01 ~ G_04 + G_12
  c(0.000, 1.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.299, 0.000, 0.000, 0.000, 0.000, 0.229, 0.000, 0.000, 0.000), # G_02 ~ G_09 + G_14
  c(0.000, 0.000, 1.000, 0.000, 0.178, 0.000, 0.208, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000), # G_03 ~ G_05 + G_07
  c(0.000, 0.211, 0.000, 1.000, 0.000, 0.000, 0.000, 0.200, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000), # G_04 ~ G_02 + G_08
  c(0.000, 0.000, 0.178, 0.000, 1.000, 0.254, 0.000, 0.000, 0.000, 0.000, 0.000, 0.256, 0.000, 0.000, 0.000, 0.000, 0.000), # G_05 ~ G_03 + G_06
  c(0.000, 0.000, 0.000, 0.000, 0.254, 1.000, 0.000, 0.000, 0.000, 0.249, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000), # G_06 ~ G_05 + G_10
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.297, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000), # G_07 ~ G_11
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.271, 0.295, 0.000, 0.000), # G_08 ~ G_14 + G_15
  c(0.000, 0.299, 0.000, 0.000, 0.166, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000), # G_09 ~ G_02 + G_05
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.249, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.219, 0.000, 0.000, 0.000), # G_10 ~ G_06 + G_14
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.297, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.180, 0.000, 0.000), # G_11 ~ G_07 + G_15
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.295, 0.000, 0.000), # G_12 ~ G_15
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000, 0.000), # G_13 ~ 
  c(0.000, 0.229, 0.000, 0.000, 0.000, 0.000, 0.000, 0.271, 0.000, 0.000, 0.000, 0.000, 0.000, 1.000, 0.000, 0.000, 0.000), # G_14 ~ G_02 + G_08
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.295, 0.000, 0.000, 1.000, 0.000, 0.000), # G_15 ~ G_12
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 1.000, 0.000), # G_16 ~ 
  c(0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 0.000, 1.000)  # G_17 ~ 
)

set.seed(1999)
groups <- paste0(rep(c("F","M"), each=4), "_", rep(c("E","B","A","O"), 2))
jM <- function(p, s = 0.02) {
  p2 <- p + rnorm(length(p), 0, s)
  p2[p2 < 0] <- 0              
  p2 <- p2 / sum(p2)
  c(p2[1], p2[1] + p2[2])
}
jC <- function(C) {
  N <- matrix(rnorm(17*17, 0, 0.02), 17, 17)
  C2 <- (C + t(C) + N + t(N)) / 4
  diag(C2) <- 1; C2[C2 < 0] <- 0; C2[C == 0] <- 0
  C2 <- (C2 + t(C2)) / 2
  C2 <- as.matrix(Matrix::nearPD(C2, corr = T)$mat)
}
genotype_marginals_by_group <- lapply(groups, function(.) lapply(genotype_marginal, jM))
genotype_correlations_by_group <- lapply(groups, function(.) jC(genotype_correlation))

save(genotype_marginals_by_group, genotype_correlations_by_group, file = "genotypes.RData")
